docs/chapter2-payload-image-format.rst

```diff
@@ -0,0 +1,671 @@
+                                              FIT without needing this property to indicate
+                                              which images are needed.
+   compatible        O           string list  List of compatible strings for Platform Init
+                                              to use when selecting the best configuration,
+                                              in order from most specific / desirable to
+                                              least. This may be used to support using the
+                                              same FIT with many different boards, where
+                                              Platform Init understands the concept of a
+                                              compatible string and can make this
+                                              determination.
+   require-fit       O           boolean      If this property exists, it means the whole
+                                              fit image shall be loaded together before
+                                              Platform Init calling payload entry.
+   ================= =========== ============ ==============================================
+
+
+2.3.1 Example FIT
+~~~~~~~~~~~~~~~~~
+
+Shown below is a FIT structure in source form (Image Tree Source) of a Tianocore
+Payload. It shows three images and a single configuration: the main image
+"tianocore" and 2 more images "uefi-fv" and "bds-fv". "uefi-fv" and "bds-fv" are
+included to be used later by the main image after the main image is executed.
+
+.. code-block:: none
+
+   / {
+    	description = “Uefi Payload”
+    	timestamp = <0x00000000>
+    	#address-cells = <0x02>;
+    	size = <0x00385000>	
+    	spec-version = <0x00000100>;
+    	build-revision = <0x00010105>;
+    	images {
+    		tianocore {
+    			description = "Tianocore edk2-stable202208";
+    			timestamp = <0x00000000>
+    			project = "tianocore";
+    			arch = "x86_64";
+    			type = "flat-binary";
+    			capabilities = "smm-rebase", "...";
+    			producer = "My company";
+    			data-offset = <...data…>;
+    			data-size = <...data…>;
+    			reloc-start = <start offset of reloc table within data>;
+    			entry-start = <0x121b10>;
+    			load = <0x120000>;
+    		};
+    		uefi-fv {    // showing how to have multiple images
+    	      description = "UEFI Firmware Volume";
+    	      type = "flat-binary";
+    			arch = "x86_64";
+    			project = "tianocore";
+    			compression = "lzma";
+    			data-offset = <...data…>;
+    			data-size = <...data…>;
+    		};
+    		bds-fv {
+    			description = "BDS Firmware Volume";
+    			type = "flat-binary";
+    			arch = "x86_64";
+    			project = "tianocore";
+    		compression = "lzma";
+    		data-offset = <...data…>;
+    		data-size = <...data…>;
+    	};
+    };
+    	configurations {
+    		default = "conf-1";
+    		conf-1 {
+    			firmware = "tianocore"
+    			require-fit;
```

Comment:
Maybe need to change it to require-fit = "TRUE", due to FDT don't have API for boolean case.

Reply:
this example is correct, as DT spec mentions:
_Properties might have an empty value if conveying true-false information. In this case, the presence or absence of the property is sufficiently descriptive_

We just need to update the property description to match it.

Reply:
Basically, 

it's unsupported syntax
- require-fit;

support syntax like below
- require-fit = "true"; <-------- treat as string
- require-fit = <0x00000001>; <-------- treat as uint32
- require-fit = <0x0000000000000001>; <-------- treat as uint64

Don't find any API can support require-fit, it's my conern.
URL: https://github.com/devicetree-org/pylibfdt/blob/main/libfdt/libfdt.i
![image](https://github.com/UniversalPayload/spec/assets/96405988/4d193297-280b-48fb-a868-73a4a86b83b2)

Reply:
@sjg20 would you have any idea on this?

Reply:
It is OK to do this although pylibfdt has no specific support for it. I will ask about adding it. You can check it with:

if fdt.getprop(node, 'require-fit', libfdt.QUIET_NOTFOUND) != -libfdt.NOTFOUND:
   print('require-fit' is present')

Reply:
See devicetree spec section 2.2.4:

"Properties might have an empty value if conveying true-false information. In this case, the presence or absence of the
property is sufficiently descriptive."

Reply:
thanks Simon! I have also updated the require-fit value type from boolean to empty to match DT spec.

Reply:
Hi, you can just use the setprop method with an empty value

Reply:
@sjg20 , from my understanding, BIOS code is fine as we can call to any function from libfdt, but just current FDT binary build steps do not support "val is NULL" and will trigger build error.

@gguo11837463, below function seems what we need, could you share what change might be needed in the  build steps? Does that mean we need to add new function "def setprop_empty" in libfdt.i to support empty value case?


libfdt\libfdt.h:
/**
 * fdt_setprop_empty - set a property to an empty value
 * @fdt: pointer to the device tree blob
 * @nodeoffset: offset of the node whose property to change
 * @name: name of the property to change
 *
 * fdt_setprop_empty() sets the value of the named property in the
 * given node to an empty (zero length) value, or creates a new empty
 * property if it does not already exist.
 *
 * This function may insert or delete data from the blob, and will
 * therefore change the offsets of some existing nodes.
 *
 * returns:
 *	0, on success
 *	-FDT_ERR_NOSPACE, there is insufficient free space in the blob to
 *		contain the new property value
 *	-FDT_ERR_BADOFFSET, nodeoffset did not point to FDT_BEGIN_NODE tag
 *	-FDT_ERR_BADLAYOUT,
 *	-FDT_ERR_BADMAGIC,
 *	-FDT_ERR_BADVERSION,
 *	-FDT_ERR_BADSTATE,
 *	-FDT_ERR_BADSTRUCTURE,
 *	-FDT_ERR_BADLAYOUT,
 *	-FDT_ERR_TRUNCATED, standard meanings
 */
#define fdt_setprop_empty(fdt, nodeoffset, name) \
	fdt_setprop((fdt), (nodeoffset), (name), NULL, 0)

Reply:
It's unlocked now

Reply:
what does 'unlocked' mean here?

Reply:
Fix now.
![image](https://github.com/UniversalPayload/spec/assets/96405988/e2768720-019e-4cff-8402-e912f897e6fa)